### ISSUE TITLE
[EMU] crash accessing "Clicky" member

### ISSUE THREAD

**Issue by aquietone** [macroquest/macroquest#651]
Recently added use of a summoned mage clicky nuke with charges to a script and toons are periodically crashing. The summoned item goes to each group member and between the 6 of them i see maybe 2-3 of these crashes throughout a day. The crash is always after using the last charge and the item goes away. I've tried to reproduce it with a loop creating and using waterstones without any luck so far.

MQ2DataAPI.cpp:276
```
if (type->GetMember(std::move(VarPtr), Member.c_str(), pIndex, Result))
```

```
Unhandled exception at 0x66256325 in eed950fb-d4ff-4c19-a7d2-dc653ec90d1a.dmp: 0xC0000005: Access violation executing location 0x66256325.
```

```
  | Name | Value | Type
-- | -- | -- | --
  | <begin>$L0 | Variable is optimized away and not available. |  
  | <end>$L0 | Variable is optimized away and not available. |  
  | checkFirst | false | bool
  | extIter | Variable is optimized away and not available. |  
▶ | Member | "Clicky" | const std::string &
▶ | pIndex | 0x54b0edde "" | char *
▶ | Result | MQTypeVar { String: 0x00000000 <NULL> } | mq::MQTypeVar &
  | result | Variable is optimized away and not available. |  
  | type | Error reading register value. |  
▶ | VarPtr | Ptr { 0x659de7ac "00{ePc{eP^{eÐo{eào{eà‘}e U{e€g}ePr\|eº}e°L{epp{eÐ5¢e`9{ePc{epa{eÐo{e n{e" } | mq::MQVarPtr &
```

```
 	66256325()	Unknown
 	[Frames below may be incorrect and/or missing]	
>	MQ2Main.dll!mq::EvaluateMacroDataMember(mq::datatypes::MQ2Type * type, mq::MQVarPtr & VarPtr, mq::MQTypeVar & Result, const std::string & Member, char * pIndex, bool checkFirst) Line 276	C++
 	MQ2Main.dll!EvaluateMacroDataMember(mq::datatypes::MQ2Type * pType, mq::MQVarPtr VarPtr, mq::MQTypeVar & Result, const char * Member, char * pIndex) Line 480	C++
 	mq2lua.DLL!mq::lua::lua_MQTypeVar::EvaluateMember(const char * index) Line 195	C++
 	mq2lua.DLL!mq::lua::lua_MQTypeVar::Get(sol::basic_object<sol::stack_reference> key, sol::this_state L) Line 271	C++
 	[Inline Frame] mq2lua.DLL!sol::member_function_wrapper<sol::basic_object<sol::basic_reference<0>> (__thiscall mq::lua::lua_MQTypeVar::*)(sol::basic_object<sol::stack_reference>,sol::this_state)const ,sol::basic_object<sol::basic_reference<0>>,mq::lua::lua_MQTypeVar,sol::basic_object<sol::stack_reference>,sol::this_state>::call(sol::basic_object<sol::basic_reference<0>>(const mq::lua::lua_MQTypeVar::*)(sol::basic_object<sol::stack_reference>, sol::this_state) &) Line 16278	C++
 	[Inline Frame] mq2lua.DLL!sol::member_function_wrapper<sol::basic_object<sol::basic_reference<0>> (__thiscall mq::lua::lua_MQTypeVar::*)(sol::basic_object<sol::stack_reference>,sol::this_state)const ,sol::basic_object<sol::basic_reference<0>>,mq::lua::lua_MQTypeVar,sol::basic_object<sol::stack_reference>,sol::this_state>::caller::operator()(sol::basic_object<sol::basic_reference<0>>(const mq::lua::lua_MQTypeVar::*)(sol::basic_object<sol::stack_reference>, sol::this_state) &) Line 16284	C++
 	[Inline Frame] mq2lua.DLL!sol::stack::stack_detail::eval(sol::types<>) Line 15028	C++
 	[Inline Frame] mq2lua.DLL!sol::stack::stack_detail::eval(sol::types<sol::this_state>) Line 15034	C++
 	[Inline Frame] mq2lua.DLL!sol::stack::stack_detail::eval(sol::types<sol::basic_object<sol::stack_reference>,sol::this_state>) Line 15034	C++
 	[Inline Frame] mq2lua.DLL!sol::stack::stack_detail::call(sol::types<sol::basic_object<sol::basic_reference<0>>>) Line 15058	C++
 	[Inline Frame] mq2lua.DLL!sol::stack::call(sol::types<sol::basic_object<sol::basic_reference<0>>>) Line 15076	C++
 	mq2lua.DLL!sol::stack::call_into_lua<0,1,sol::basic_object<sol::basic_reference<0>>,sol::basic_object<sol::stack_reference>,sol::this_state,sol::member_function_wrapper<sol::basic_object<sol::basic_reference<0>> (__thiscall mq::lua::lua_MQTypeVar::*)(sol::basic_object<sol::stack_reference>,sol::this_state)const ,sol::basic_object<sol::basic_reference<0>>,mq::lua::lua_MQTypeVar,sol::basic_object<sol::stack_reference>,sol::this_state>::caller,sol::basic_object<sol::basic_reference<0>> (__thiscall mq::lua::lua_MQTypeVar::*&)(sol::basic_object<sol::stack_reference>,sol::this_state)const ,mq::lua::lua_MQTypeVar &>(sol::types<sol::basic_object<sol::basic_reference<0>>> tr, sol::types<sol::basic_object<sol::stack_reference>,sol::this_state> ta, lua_State * L, int start, sol::member_function_wrapper<sol::basic_object<sol::basic_reference<0>> (__thiscall mq::lua::lua_MQTypeVar::*)(sol::basic_object<sol::stack_reference>,sol::this_state)const ,sol::basic_object<sol::basic_reference<0>>,mq::lua::lua_MQTypeVar,sol::basic_object<sol::stack_reference>,sol::this_state>::caller && fx, sol::basic_object<sol::basic_reference<0>>(const mq::lua::lua_MQTypeVar::*)(sol::basic_object<sol::stack_reference>, sol::this_state) & <fxargs_0>, mq::lua::lua_MQTypeVar & <fxargs_1>) Line 15128	C++
 	[Inline Frame] mq2lua.DLL!sol::call_detail::lua_call_wrapper<mq::lua::lua_MQTypeVar,sol::basic_object<sol::basic_reference<0>> (__thiscall mq::lua::lua_MQTypeVar::*)(sol::basic_object<sol::stack_reference>,sol::this_state)const ,1,0,0,0,1,void>::call(lua_State *) Line 17027	C++
 	[Inline Frame] mq2lua.DLL!sol::call_detail::lua_call_wrapper<mq::lua::lua_MQTypeVar,sol::basic_object<sol::basic_reference<0>> (__thiscall mq::lua::lua_MQTypeVar::*)(sol::basic_object<sol::stack_reference>,sol::this_state)const ,1,0,0,0,1,void>::call(lua_State *) Line 17020	C++
 	[Inline Frame] mq2lua.DLL!sol::call_detail::call_wrapped(lua_State *) Line 17418	C++
 	mq2lua.DLL!sol::u_detail::binding<enum sol::meta_function,sol::basic_object<sol::basic_reference<0>> (__thiscall mq::lua::lua_MQTypeVar::*)(sol::basic_object<sol::stack_reference>,sol::this_state)const ,mq::lua::lua_MQTypeVar>::call_with_<1,0>(lua_State * L, void * target) Line 21772	C++
 	mq2lua.DLL!sol::u_detail::usertype_storage_base::self_index_call<0,0,0>(sol::types<> __formal, lua_State * L, sol::u_detail::usertype_storage_base & self) Line 22231	C++
 	mq2lua.DLL!sol::u_detail::usertype_storage<mq::lua::lua_MQTypeVar>::index_call_<0,0>(lua_State * L) Line 22280	C++
 	mq2lua.DLL!sol::detail::lua_cfunction_trampoline(lua_State * L, int(*)(lua_State *) f) Line 8008	C++
 	[Inline Frame] mq2lua.DLL!sol::detail::static_trampoline(lua_State *) Line 8012	C++
 	mq2lua.DLL!sol::u_detail::usertype_storage<mq::lua::lua_MQTypeVar>::index_call<0>(lua_State * L) Line 22292	C++
 	mq2lua.DLL!sol::detail::trampoline<int (__cdecl*&)(lua_State *)>(lua_State * L, int(*)(lua_State *) & f) Line 8059	C++
 	mq2lua.DLL!sol::detail::c_trampoline(lua_State * L, int(*)(lua_State *) f) Line 8062	C++
 	[External Code]	
```

The code is something like
```
        local clickyItem = mq.TLO.FindItem('='..clicky)
        if clickyItem() and clickyItem.Timer() == '0' then
            local castTime = clickyItem.Clicky.Spell.CastTime()
            mq.cmdf('/useitem "%s"', clickyItem.Name())
            mq.delay(50+castTime, function() return not mq.TLO.Me.Casting() end)
        end
```

**Comment by brainiac** [comment]
the item is going out of scope when it is consumed, because not all parts of it are reference counted. takes some work, I added support for the reference counting but being able to hold the reference while utilizing tan itemspell is a bit trickier with the macro system

**Comment by brainiac** [comment]
fixed via 6881fb0e2ded66cffe7654711264533603a7d11a
